rfcs/20231005-hybrid-quantization.md

```diff
@@ -0,0 +1,158 @@
+
+* It aligns well with the existing behavior of the corresponding HLO ops which
+allow operand types to differ and upcast operands to the highest-precision type
+in such a case.
+* Being able to express weight-only natively in StableHLO, it does not rely on
+pattern matching by backend to avoid patterns that might result in bad
+performance.
+* Constant and dequantize can unintentionally be folded in downstreams and
+hybrid op removes concern on constant folding. We can also consider inserting
+optimization_barrier in between to prevent constant folding, but this requires
+upstream frameworks to embed optimization_barrier and downstreams to pattern
+match additional optimization_barrier. It is hard to guarantee these
+requirements on use-cases across various frameworks and hardwares.
+
+## Examples
+
+Here are examples of hybrid quantized convolution and dot_general. Hybrid
+quantized ops will get float input(or lhs) and quantized weight(or rhs) as
+operands and output float results.
+
+```mlir
+%conv = "stablehlo.convolution"(%input_f, %weight_q) ... : (tensor<...xf32>, !quant.uniform<i8:f32, scale:zp>) -> tensor<...xf32>>
+```
+
+```mlir
+%dot = "stablehlo.dot_general"(%lhs_f, %rhs_q) ... : (tensor<...xf32>, !quant.uniform<i8:f32, scale:zp>) -> tensor<...xf32>>
+```
+
+## Proposed spec changes
+
+We propose to modify operation semantics and a few constraints on quantized
+convolution and dot_general to represent weight-only quantization using hybrid
+op.
+
+### hybrid_dequantize_then_op semantics
+
+We propose to define `hybrid_dequantize_then_op` semantics as part of quantization
+computations. 
+
+* `hybrid_dequantize_then_op` is used to specify weight-only quantization for
+hybrid op which accepts lhs in floating-point and rhs in quantized types. It
+dequantizes quantized inputs into their expressed types and performs computation
+in float. Element type of float lhs tensor and expressed type of quantized rhs
+tensor should be identical.
+
+```python
+def hybrid_dequantize_then_op(op, lhs, rhs):
+  assert(is_float(lhs) and is_quantized(rhs) and element_type(lhs) == expressed_type(rhs))
+  return op(lhs, dequantize(rhs))
+```
+
+### convolution
+
+#### Operation semantics for hybrid op
+
+For hybrid quantized types, performs `hybrid_dequantize_then_op( lambda lhs,
+rhs: convolution(lhs, rhs, window_strides, padding, lhs_dilation, rhs_dilation,
+window_reversal, input_batch_dimension, input_feature_dimension,
+input_spatial_dimensions, kernel_input_feature_dimension,
+kernel_output_feature_dimension, kernel_spatial_dimensions,
+output_batch_dimension, output_feature_dimension, output_spatial_dimensions,
+feature_group_count, batch_group_count, precision_config), lhs, rhs)`.
+
+#### Current constraints
+
+* If the operation uses quantized tensors:
+  * (C28) `is_quantized_tensor(lhs) and is_quantized_tensor(rhs) and
+    is_quantized_tensor(result)`.
+  * (C29) `storage_type(lhs) =  storage_type(rhs)`.
```

Comment:
` = storage_type(result)` ?

Reply:
Realizing this is current constraints, but this also shows up in the new constraints to same question

Reply:
Followed up with @sdasgup3 - result storage type can be larger.